R/expect_setequal_linter.R

```diff
@@ -0,0 +1,39 @@
+#' Require usage of expect_setequal() over expect_identical(sort(...), ...)
+#'
+#' `expect_setequal()` is designed for testing an output, regardless of order;
+#'   for example, this is particularly useful for SQL, which is typically
+#'   row-order-agnostic.
+#'
+#' Note that, in the presence of possible duplicates,
+#'   `expect_identical(sort(x), y)` won't be the same as
+#'   `expect_setequal(x, y)`. This linter encourages a separate test for
+#'   duplicates rather than integrating two tests into one.
+#'
+#' @evalRd rd_tags("expect_setequal_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+expect_setequal_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    xpath <- glue::glue("//expr[
+      expr[SYMBOL_FUNCTION_CALL[text() = 'expect_identical' or text() = 'expect_equal']]
+      and expr[expr[SYMBOL_FUNCTION_CALL[text() = 'sort']]]
```

Comment:
`sort()` doesn't deduplicate, so we can only ascertain the lint for `sort(unique())` or `unique(sort())`.

Reply:
Our sense is that people are using `expect_equal(sort())` in cases without dups, hence pointing out in the lint message that dup tests are different. Even with dups, `expect_equal(sort(x), sort(y))` smells fishy to me -- a bit hard to imagine when such a test would not be better written in another way.

This linter only caught a handful of cases in practice as is. Maybe we should just exclude this one?

Reply:
Fine for me.
I can imagine `expect_equal(sort(..), ...)` to be useful when testing some join like functionality.

Reply:
Makes sense... I wonder if `expect_setequal()` is maybe itself undesirable -- a better test should probably specify exactly how the sets are equal, right? Maybe `expect_setequal()` should be replaced with `expect_identical(sort(unique(x)), y)` 🤔

Anyway, closing for now.